### PR TITLE
Capture the number of unique IPs accessing GOCDB via the load balancers

### PR DESCRIPTION
**Pull request by gregcorbett**
- as this is a metric of interest to EOSC Hub

### CODE REVIEW

File: metrics_gocdb.py [hunk 1]
```diff
@@ -172,6 +172,72 @@ def get_queries():
...
 
+
+def get_unique_ips_over_28_days():
+    """
+    Get the number of unique IP accessing GOCDB over the last 28 days.
+
+    Only accesses via the load balancers will be counted.
+    """
+    date = datetime.now()
+    # Use a set to leverage the fact duplicates are handled by the data
```
Comment (tofu-rocketry) [inline metrics_gocdb.py:182]: 👍

File: metrics_gocdb.py [hunk 2]
```diff
@@ -172,6 +172,72 @@ def get_queries():
...
+    # structure.
+    unique_ips = set()
+    # For the last 28 days (not counting today), query the unique IPs per day.
+    for _ in range(0, 28):
```
Comment (tofu-rocketry) [inline metrics_gocdb.py]: Took me a few moments to figure out how this loop works. Kinda feel that `date` should be nearer to where it's used, but not sure.

File: metrics_gocdb.py [hunk 3]
```diff
@@ -172,6 +172,72 @@ def get_queries():
...
+        date = date - timedelta(1)
```
Comment (tofu-rocketry) [inline metrics_gocdb.py]: Can be more explicit:
```suggestion
        date = date - timedelta(days=1)
```

Comment (gregcorbett) [reply metrics_gocdb.py]: +1

File: metrics_gocdb.py (hunk 2 above)
Comment (gregcorbett) [reply metrics_gocdb.py]: agreed, I've pushed a new commit addressing this. Does it make the loop clearer?

Comment (tofu-rocketry) [reply metrics_gocdb.py]: I actually meant where it's used on line 229, but this does help.

**Review by tofu-rocketry — APPROVED** [review]

File: metrics_gocdb.py [hunk 4]
```diff
@@ -172,6 +172,72 @@ def get_queries():
...
+
+    Only accesses via the load balancers will be counted.
+    """
+    # Use a set to leverage the fact duplicates are handled by the data
+    # structure.
+    unique_ips = set()
+    # For the last 28 days (not counting today), query the unique IPs per day.
+    date = datetime.now()
+    for _ in range(0, 28):
+        date = date - timedelta(days=1)
```
Comment (jrha) [inline metrics_gocdb.py]: I agree with @tofu-rocketry and this is still hard to follow, why not do something like:

```python
today = datetime.now()
for i in range(0, 28):
    date = today - timedelta(days=i)
    …
```

Comment (tofu-rocketry) [reply metrics_gocdb.py]: He's stepping through the dates in the loop, so has to do an in-place subtraction.

As an aside, do you know about the in-place operators, @gregcorbett? `+=, -=, *=` etc. https://docs.python.org/3/reference/simple_stmts.html#augmented-assignment-statements

Comment (gregcorbett) [reply metrics_gocdb.py]: Can I not do as @jrha sugest @tofu-rocketry? I'd have to fuss with the loop start a bit to have it loop over the same days `(today - timedelta(days=0))` would equal `today`....

I could also move defining `ELASTICSEARCH` and the `params_dict` out of the loop (as they are constants) to make the loop smaller and easier to follow?

Comment (tofu-rocketry) [reply metrics_gocdb.py]: Ah, I see, you'd always calculate `date` in  relation to today. Yes, that seems better. I'd not noticed the change to using `i`.

Comment (gregcorbett) [reply metrics_gocdb.py]: change made

**Review by jrha — APPROVED** [review]
